[Issue #21] Adds the ability to put count and ratio in the layout.

hfytr opened the pull request:
This PR adds a --layout-format flag which enables setting the layout to an arbitrary string containing count {c}, ratio {r}, and layout image {l}. for example passing --layout-format "{c} {r}\n{l}" may result in something like:
```
3 0.5
│├──┤
```
This flag is optional, and takes the default value {l}, which should have the same behaviour as before this PR. Within the string, brackets are allowed using \\{, \\}. \n, \r, \t, \v, and nerdfont icons are also supported.

Tests have not been written, and I am definitely not a C programmer, so I would be more surprised than not if the code has no vulnerabilities / edge cases that I am missing.

NOTE 1: I committed the flake.lock / flake.nix for future contributors, but its up to you if you want to keep them.

NOTE 2: There is no clang-format, so my formatting is likely wrong.

Commit:
Works, still need \n and emojis;
EDIT: emojis "just work." I'll take it.
EDIT2: everything good now

Commit:
lint / check includes / updated help

alex-courtis commented:
This looks great - very easy to define.

I'll test and review tomorrow.

alex-courtis commented:
Happy to help out with tests; they should be straightforward unit.

hfytr commented:
Hello, it seems that I ran make test, and then make iwyu. Incorporating the suggestions of make iwyu broke the tests. Anyway, should be good now.

Commit:
linting

alex-courtis commented:
> Hello, it seems that I ran make test, and then make iwyu. (thoughtlessly) Incorporating the suggestions of make iwyu broke the tests. Anyway, should be good now.

Inconsistent iwyu is becoming a problem, I think I'll just remove it.

I've just made a config change #23 to cppcheck, for versions greater than 2.7 (CI)

If it's OK with you I'll:
* update this branch
* write and push tests

hfytr commented:
Looks good! Thanks for putting up with me not following CONTRIBUTING.md. For future, how do I run checks locally?

alex-courtis commented:
> Looks good! Thanks for putting up with me not following CONTRIBUTING.md. For future, how do I run checks locally?

You're all good. It looks like you've been running all of them. `cppcheck` is the useful one: https://github.com/alex-courtis/wideriver/blob/master/CONTRIBUTING.md#lint

Not sure how your environment's setup, but you can manually check cross-compiler:
```sh
make CC=gcc clean all
make CC=clang clean all
```

I'll deal with iwyu

alex-courtis committed:
Merge branch 'master' into master

alex-courtis committed:
add iwyu mappings

alex-courtis commented on src/layout.c:
```diff
@@ -54,6 +56,88 @@ const char *description_info(const struct Demand* const demand, const struct Tag
+	return desc;
+}
+
 const char *description_debug(const struct Demand* const demand, const struct Tag* const tag) {
```
We can delete this now :)

alex-courtis committed:
add layout test and add layout-format to args test

alex-courtis committed:
terminate static layout desc string

alex-courtis commented on inc/enum.h:
```diff
@@ -27,6 +27,14 @@ enum LogThreshold {
+enum LayoutFormatElem {
+	LAYOUT = 'l',
+	COUNT = 'c',
+	RATIO = 'r',
```
Can we please add `{n}` for name i.e. wide, left etc. ?

This will allow users to extract the current layout, as the image is variable.

alex-courtis committed:
add default layout to usage

alex-courtis reviewed (changes requested):
Looking good, I'm using it right now.

I've added a test suite for layout and added one basic test for `cfg_set_layout_format`.

Please

- [x] add `{n}`
- [x] delete `description_debug` and move `description_info` into `layout_description`
- [ ] complete test cases for `cfg_set_layout_format`

I'll think about how we can simplify `description_info`. I've seen something similar but I can't recall it right now.

I'm unavailable until next weekend.

alex-courtis commented:
> NOTE 2: There is no clang-format, so my formatting is likely wrong.

Sorry, I have to get to that...

What IDE do you use? vim `gg=G` is fine.

Commit:
added {n} for name (left/right/monocle/etc)

Commit:
string vulnerability things

alex-courtis commented:
We could use a simple string replacer:

```c
	static char desc[LAYOUT_FORMAT_LEN+5];

	snprintf(sratio, 10, "%g", ratio);
	snprintf(scount, 10, "%d", count);

	char *counted = replace_string(cfg->layout_format, "{c}", scount);

	char *ratioed = replace_string(counted, "{r}", sratio);
	free(counted);

	char *imaged = replace_string(ratioed, "{l}", image);
	free(ratioed);

	strncpy(desc, imaged, LAYOUT_FORMAT_LEN + 5);
	free(imaged);

	return desc;
```

```c
// return a copy of str with instances of target substituted with replacement
// call frees
char *replace_string(const char * const str, const char * const target, const char * const replacement) {
	char *match = strstr(str, target);
	if (!match) {
		return strdup(str);
	}

	// allocate result
	size_t len = strlen(str) + strlen(replacement) - strlen(target) + 1;
	char *res = calloc(len, sizeof(char));

	// find string up to match
	size_t len_pre = match - str + 1;
	char pre[len_pre];
	snprintf(pre, len_pre, "%s", str);

	// format
	snprintf(res, len, "%s%s%s", pre, replacement, match + strlen(target));

	// recurse
	char *next = replace_string(res, target, replacement);
	free(res);
	return next;
}
```

hfytr committed:
merged

hfytr committed:
merged

hfytr commented:
OK, so I added {n}, changed some things to use strn functions. my impl of {n} is pretty bad, as I wasn't sure how to integrate with the existing switch / case. Also, I/you will have to change it to use your string replacer (which LGTM).

hfytr committed:
merge
Merge remote-tracking branch 'refs/remotes/origin/master'

alex-courtis committed:
add string_replace

alex-courtis committed:
layout_description uses string_replace

alex-courtis committed:
layout_description uses string_replace

alex-courtis committed:
layout_description and image are caller frees

alex-courtis committed:
temporarily disable gcc due to false positive stringop-truncation

alex-courtis commented:
> OK, so I added {n}, changed some things to use strn functions. my impl of {n} is pretty bad, as I wasn't sure how to integrate with the existing switch / case. Also, I/you will have to change it to use your string replacer (which LGTM).

Looks good, moved it to the very hidden function from enum.c

> EDIT: Your tests were using (the now nonexistent) description_info, so i switched them to layout_description() instead.

Updated layout_description to use string replace. Changed the layout functions to be caller frees, as per the original comments.

> Sorry for the horrible git history.

Don't worry about git history, it will be squashed on merge.

We do need to take care of the documentation - man and readme. I'd be grateful if you could update `doc/templ/40.man.readme.md` with the new options.
Don't worry if you can't run `make doc`, it's pretty hacky.

alex-courtis committed:
temporarily use clang for test-vg due to false positive stringop-truncation

alex-courtis commented:
Unfortunately I'm out of time; I'll get back to this next weekend.

This one will be a slow burn...

alex-courtis commented on src/usage.c:
```diff
@@ -17,59 +17,77 @@ void usage(const int status) {
 			"\n"
-			"  --count-master                  count                                 %d           %d <= count\n"
-			"  --ratio-master                  ratio                                 %.02f      %.1g <= ratio <= %.1g\n"
+			"  --count-master                  count                             "
```
Please don't change formatting for existing lines.

alex-courtis committed:
Revert "temporarily use clang for test-vg due to false positive stringop-truncation"

This reverts commit 136b995987404c91833ce4fe42730802a977f7f7.

alex-courtis commented on src/cfg.c:
```diff
@@ -190,6 +191,39 @@ bool cfg_set_border_width_monocle(const char *s) {
 	}
 }
 
+bool cfg_set_layout_format(const char *s) {
```
I don't quite understand the failure cases of this function; I might be missing the point, but could it just be "garbage in, garbage out"?

Can you please add tests to `tst-args_cli.c` to cover these cases?

alex-courtis committed:
Revert "temporarily disable gcc due to false positive stringop-truncation"

This reverts commit bf749b98276a182bf58f01368de63e659e07f4cf.

alex-courtis committed:
use memcpy (recommended) instead of stpncpy to avoid spurious gcc warnings

alex-courtis committed:
Merge branch 'master' into master

nirodhvana commented:
@alex-courtis will this be merged?

alex-courtis commented:
> @alex-courtis will this be merged?

I'm not quite sure of the status of this PR. There is still a lot of unnecessary code such as
* all the `\t` `\r` escaping
* `--layout-format` parsing
  * does not seem to be necessary at all
  * doesn't seem safe from a memory or security perspective

I attempted a rebase and a merge from upstream - ~1 year ago - however I gave up.

The best course of action might be to create a fresh branch and pull in only the minimum required.

As usual, pull requests are most gratefully appreciated, see [CONTRIBUTING.md](https://github.com/alex-courtis/wideriver/blob/master/CONTRIBUTING.md)
